Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+                if (fieldType == typeof(byte[]))
+                {
+                    return SFDataConverter.HexToBytes(value);
+                }
+                throw new Exception($"Could not read BINARY into {fieldType}");
+            }
+            if (IsObjectMetadata(fieldMetadata)) // semi structured object
+            {
+                return json.ToString();
+            }
+            if (IsArrayMetadata(fieldMetadata)) // semi structured array
+            {
+                return json.ToString();
+            }
+            if (IsVariantMetadata(fieldMetadata))
+            {
+                return json.ToString();
+            }
+            throw new Exception($"Could not read {fieldMetadata.type} into {fieldType}");
+        }
+
+        private static object ConvertToArray(Type type, Type elementType, List<FieldMetadata> fields, JToken json)
+        {
+            if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+            {
+                return null;
+            }
+            if (json.Type != JTokenType.Array)
+            {
+                throw new Exception("Json is not an array");
+            }
+            var jsonArray = (JArray)json;
+            var arrayType = MakeArrayType(type, elementType);
+            var result = (IList) Activator.CreateInstance(arrayType, jsonArray.Count);
+            var elementMetadata = fields[0];
+            for (var i = 0; i < jsonArray.Count; i++)
+            {
+                result[i] = ConvertToStructuredOrUnstructuredValue(elementType, elementMetadata, jsonArray[i]);
+            }
+            if (type != arrayType)
+            {
+                var listType = type.IsAbstract ? typeof(List<>).MakeGenericType(elementType) : type;
+                var list = (IList) Activator.CreateInstance(listType);
+                for (int i = 0; i < result.Count; i++)
+                {
+                    list.Add(result[i]);
+                }
+                return list;
+            }
+
+            return result;
+        }
+
+        private static object ConvertToMap(Type type, Type keyType, Type valueType, List<FieldMetadata> fields, JToken json)
+        {
+            if (keyType != typeof(string)
+                && keyType != typeof(int) && keyType != typeof(int?)
+                && keyType != typeof(long) && keyType != typeof(long?))
+            {
+                throw new Exception($"Unsupported key type of dictionary {keyType}");
+            }
+            if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+            {
+                return null;
+            }
+            if (json.Type != JTokenType.Object)
+            {
+                throw new Exception("Json is not an object");
+            }
+            if (fields == null || fields.Count != 2)
+            {
+                throw new Exception("Expected map to have 2 metadata fields");
```

Comment:
what does this map contain? and what happened? and maybe "should have" instead of "to have"

Conversion to map failed because...

Reply:
I changed to: Extracting a map failed. Map metadata should have 2 metadata fields